File: Numerical-Differentiation-and-Integration/GaussianQuadrature.f90
program gauss_quadrature
    implicit none
    integer :: g
    real(8) :: a, b, result

    ! Input the lower and upper limits
    print *, "Enter the lower limit: "
    read *, a
    print *, "Enter the upper limit: "
    read *, b

    ! Input the number of points for the Gaussian formula
    print *, "Number of points to use in Gauss point-formula (2, 3, 4, 5): "
    read *, g

    select case (g)
        case (2)
            result = gauss_2_point(a, b)
            print *, "Result using Gauss 2-point formula is: ", result
        case (3)
            result = gauss_3_point(a, b)
            print *, "Result using Gauss 3-point formula is: ", result
        case (4)
            result = gauss_4_point(a, b)
            print *, "Result using Gauss 4-point formula is: ", result
        case (5)
            result = gauss_5_point(a, b)
            print *, "Result using Gauss 5-point formula is: ", result
        case default
            print *, "Formula not available."
    end select

contains

    ! Define the function to integrate (hardcoded for now)
    real(8) function f(x)
        real(8), intent(in) :: x
        f = exp(-(x**2))  ! Example function; modify as needed
    end function f

    ! Gaussian 2-point formula
    real(8) function gauss_2_point(a, b)
        real(8), intent(in) :: a, b
        real(8) :: w1, w2, x1, x2, F1, F2

        w1 = 1.0d0
        w2 = 1.0d0
        x1 = 1.0d0 / sqrt(3.0d0)
        x2 = -1.0d0 / sqrt(3.0d0)
        F1 = f(((b - a) / 2.0d0) * x1 + (b + a) / 2.0d0)
        F2 = f(((b - a) / 2.0d0) * x2 + (b + a) / 2.0d0)
        gauss_2_point = ((b - a) / 2.0d0) * (w1 * F1 + w2 * F2)
    end function gauss_2_point

    ! Gaussian 3-point formula
    real(8) function gauss_3_point(a, b)
        real(8), intent(in) :: a, b
        real(8) :: w1, w2, w3, x1, x2, x3, F1, F2, F3

        w1 = 5.0d0 / 9.0d0
        w2 = 8.0d0 / 9.0d0
        w3 = 5.0d0 / 9.0d0
        x1 = sqrt(3.0d0 / 5.0d0)
        x2 = 0.0d0
        x3 = -sqrt(3.0d0 / 5.0d0)
        F1 = f(((b - a) / 2.0d0) * x1 + (b + a) / 2.0d0)
        F2 = f(((b - a) / 2.0d0) * x2 + (b + a) / 2.0d0)
        F3 = f(((b - a) / 2.0d0) * x3 + (b + a) / 2.0d0)
        gauss_3_point = ((b - a) / 2.0d0) * (w1 * F1 + w2 * F2 + w3 * F3)
    end function gauss_3_point

    ! Gaussian 4-point formula
    real(8) function gauss_4_point(a, b)
        real(8), intent(in) :: a, b
        real(8) :: w1, w2, w3, w4, x1, x2, x3, x4, F1, F2, F3, F4

        w1 = (18.0d0 + sqrt(30.0d0)) / 36.0d0
        w2 = w1
        w3 = (18.0d0 - sqrt(30.0d0)) / 36.0d0
        w4 = w3
        x1 = sqrt(3.0d0 / 7.0d0 - (2.0d0 / 7.0d0) * sqrt(6.0d0 / 5.0d0))
        x2 = -x1
        x3 = sqrt(3.0d0 / 7.0d0 + (2.0d0 / 7.0d0) * sqrt(6.0d0 / 5.0d0))
        x4 = -x3
        F1 = f(((b - a) / 2.0d0) * x1 + (b + a) / 2.0d0)
        F2 = f(((b - a) / 2.0d0) * x2 + (b + a) / 2.0d0)
        F3 = f(((b - a) / 2.0d0) * x3 + (b + a) / 2.0d0)
        F4 = f(((b - a) / 2.0d0) * x4 + (b + a) / 2.0d0)
        gauss_4_point = ((b - a) / 2.0d0) * (w1 * F1 + w2 * F2 + w3 * F3 + w4 * F4)
    end function gauss_4_point

    ! Gaussian 5-point formula
    real(8) function gauss_5_point(a, b)
        real(8), intent(in) :: a, b
        real(8) :: w1, w2, w3, w4, w5, x1, x2, x3, x4, x5, F1, F2, F3, F4, F5

        w1 = (322.0d0 + 13.0d0 * sqrt(70.0d0)) / 900.0d0
        w2 = w1
        w3 = 128.0d0 / 225.0d0
        w4 = (322.0d0 - 13.0d0 * sqrt(70.0d0)) / 900.0d0
        w5 = w4
        x1 = (1.0d0 / 3.0d0) * sqrt(5.0d0 - 2.0d0 * sqrt(10.0d0 / 7.0d0))
        x2 = -x1
        x3 = 0.0d0
        x4 = (1.0d0 / 3.0d0) * sqrt(5.0d0 + 2.0d0 * sqrt(10.0d0 / 7.0d0))
        x5 = -x4
        F1 = f(((b - a) / 2.0d0) * x1 + (b + a) / 2.0d0)
        F2 = f(((b - a) / 2.0d0) * x2 + (b + a) / 2.0d0)
        F3 = f(((b - a) / 2.0d0) * x3 + (b + a) / 2.0d0)
        F4 = f(((b - a) / 2.0d0) * x4 + (b + a) / 2.0d0)
        F5 = f(((b - a) / 2.0d0) * x5 + (b + a) / 2.0d0)
        gauss_5_point = ((b - a) / 2.0d0) * (w1 * F1 + w2 * F2 + w3 * F3 + w4 * F4 + w5 * F5)
    end function gauss_5_point

end program gauss_quadrature

File: Numerical-Differentiation-and-Integration/GaussianQuadrature.m
f = input('Enter your function: ');
a = input('Lower limit: '); 
b = input('Upper limit: '); 
g = input('Number of points being used in Gauss point-formula (Enter 0 if you want all cases): ');
F = @(t) f(((b-a)*t + (b+a)) / 2); 

switch g 
    case 2
        w1 = 1; w2 = 1; 
        x1 = 1/sqrt(3); x2 = -1/sqrt(3); 
        G2 = ((b-a)/2) * (w1*F(x1) + w2*F(x2));
        fprintf('Result using Gauss 2-point formula is: %.5f\n', G2)
    case 3
        w1 = 5/9; w2 = 8/9; w3 = 5/9; 
        x1 = sqrt(3/5); x2 = 0; x3 = -sqrt(3/5); 
        G3 = ((b-a)/2) * (w1*F(x1) + w2*F(x2) + w3*F(x3));
        fprintf('Result using Gauss 3-point formula is: %.5f\n', G3)
    case 4
        w1 = (18 + sqrt(30)) / 36; w2 = (18 + sqrt(30)) / 36; w3 = (18 - sqrt(30)) / 36; w4 = (18 - sqrt(30)) / 36;
        x1 = sqrt(3/7 - (2/7)*(sqrt(6/5))); x2 = -sqrt(3/7 - (2/7)*(sqrt(6/5))); x3 = sqrt(3/7 + (2/7)*(sqrt(6/5))); x4 = -sqrt(3/7 + (2/7)*(sqrt(6/5)));
        G4 = ((b-a)/2) * (w1*F(x1) + w2*F(x2) + w3*F(x3 + w4*F(x4)));
        fprintf('Result using Gauss 4-point formula is: %.5f\n', G4)
    case 5
        w1 = (322+13*sqrt(70))/900; w2 = (322+13*sqrt(70))/900; w3 = 128/225; w4 = (322-13*sqrt(70))/900; w5 = (322-13*sqrt(70))/900;
        x1 = (1/3)*sqrt(5-2*sqrt(10/7)); x2 = -(1/3)*sqrt(5-2*sqrt(10/7)); x3 = 0; x4 = (1/3)*sqrt(5+2*sqrt(10/7)); x5 = -(1/3)*sqrt(5+2*sqrt(10/7));
        G5 = ((b-a)/2) * (w1*F(x1) + w2*F(x2) + w3*F(x3) + w4*F(x4) + w5*F(x5));
        fprintf('Result using Gauss 5-point formula is: %.5f\n', G5)
    otherwise 
        warning('Formula not available.')
end 